chain/src/chain.rs

```diff
@@ -693,6 +682,17 @@ impl Chain {
 			"chain: txhashset_write: finished committing the batch (head etc.)"
 		);
 
+		// Replace the chain txhashset with the newly built one.
+		{
```

Comment:
The concern here is that we have a (small) period of time between committing the batch and swapping the txhashset out for the new one.
The db/index is no longer consistent with the old txhashset at this point.
So if any other threads reads from the db we are at risk of reading inconsistent data.

I'm not sure how much of a problem this is in practice - need to think about it a bit more.

Reply:
Or maybe move to exact front of `batch.commit()` ?  more safe?

https://github.com/mimblewimble/grin/blob/7b0c8d36fc9063aec82aef220aa73f1dc460a23f/chain/src/chain.rs#L678

Reply:
There's still a non-atomic operation happening here, still a period when batch is committed but txhashet is not yet updated.

Reply:
You want a mutex to link both txhashset write and batch commit ?